kore/src/Kore/JsonRpc.hs

```diff
@@ -455,54 +458,113 @@ respond serverState moduleName runSMT =
-                            case Map.lookup (coerce name) indexedModules' of
-                                Nothing -> pure $ Left $ backendError CouldNotFindModule name
-                                Just mainModule -> do
-                                    let metadataTools = MetadataTools.build mainModule
-                                        lemmas = getSMTLemmas mainModule
-                                    serializedModule' <-
-                                        liftIO
-                                            . runSMT metadataTools lemmas
-                                            $ Exec.makeSerializedModule mainModule
-                                    internedTextCache <- liftIO $ readIORef globalInternedTextCache
-
-                                    liftIO . MVar.modifyMVar_ serverState $
-                                        \ServerState{serializedModules} -> do
-                                            let serializedDefinition =
-                                                    SerializedDefinition
-                                                        { serializedModule = serializedModule'
-                                                        , locations = kFileLocations
-                                                        , internedTextCache
-                                                        , lemmas
-                                                        }
-                                                loadedDefinition =
-                                                    LoadedDefinition
-                                                        { indexedModules = indexedModules'
-                                                        , definedNames = definedNames'
-                                                        , kFileLocations
-                                                        }
-                                            pure
-                                                ServerState
-                                                    { serializedModules =
-                                                        Map.insert (coerce name) serializedDefinition serializedModules
-                                                    , loadedDefinition
-                                                    }
-
-                                    pure . Right . AddModule $ AddModuleResult (getModuleName name)
+        AddModule AddModuleRequest{_module, nameAsId = nameAsId'} -> runExceptT $ do
+            let nameAsId = fromMaybe False nameAsId'
+            parsedModule@Module{moduleName = name} <-
+                withExceptT (backendError CouldNotParsePattern) $
+                    liftEither $
+                        parseKoreModule "<add-module>" _module
+            LoadedDefinition{indexedModules, definedNames, kFileLocations} <-
+                liftIO $ loadedDefinition <$> MVar.readMVar serverState
```

Comment:
Should maybe `takeMVar` here and `putMVar` instead of `modifyMVar` at the end (enforcing sequential consistency). Otherwise two concurrent requests could pass the check for duplicate module names but then add the same module name to the map.